From c83c5cf0f81d0090978b7e71040f5e3caa10899b Mon Sep 17 00:00:00 2001
From: Jayanthi Annadurai <jannadurai@marvell.com>
Date: Tue, 15 Jun 2021 15:03:35 -0700
Subject: [PATCH 1588/1921] drivers: marvell: rpram: Check for cn10k platform

Check the platform type is cn10k in driver init.

Change-Id: I9f58960e6ce81bc3463d2da9f24b179566224618
Signed-off-by: Jayanthi Annadurai <jannadurai@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/54168
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Reviewed-by: Sunil Kovvuri Goutham <sgoutham@marvell.com>
Reviewed-by: Chandrakala Chavva <cchavva@marvell.com>
[WK: The original patch got from Marvell sdk11.21.09]
Signed-off-by: Wenlin Kang <wenlin.kang@windriver.com>
---
 drivers/soc/marvell/cn10k-rpram.c | 31 +++++++++++++++++++++++++++++++
 1 file changed, 31 insertions(+)

diff --git a/drivers/soc/marvell/cn10k-rpram.c b/drivers/soc/marvell/cn10k-rpram.c
index a386ea42e652..0407e1fe837d 100644
--- a/drivers/soc/marvell/cn10k-rpram.c
+++ b/drivers/soc/marvell/cn10k-rpram.c
@@ -25,6 +25,8 @@
 /* Region size must be multiples of 16 MB */
 #define PRESERVE_MEMSZ_ALIGN		16
 
+/* SMC function id to check the platform type is CN10K */
+#define ARM_SMC_SVC_UID				0xc200ff01
 /* SMC function id to update persistent memory */
 #define PLAT_OCTEONTX_PERSIST_DATA_COMMAND	0xc2000b0d
 /* Arg 0: UPDATE_USERDEF_PRESERVE_MEMSZ, Update user defined
@@ -39,6 +41,27 @@ static u64 current_rpram_base;
 static struct dentry *preserve_mem_root;
 static const size_t len = PAGE_SIZE;
 
+/* This is expected CN10k response for SVC UID command */
+static const int octeontx_svc_uuid[] = {
+	0x6ff498cf,
+	0x5a4e9cfa,
+	0x2f2a3aa4,
+	0x5945b105,
+};
+
+static int check_marvell_soc_cn10k(void)
+{
+	struct arm_smccc_res res;
+
+	/* Is the other side the CN10k? */
+	arm_smccc_smc(ARM_SMC_SVC_UID, 0, 0, 0, 0, 0, 0, 0, &res);
+	if (res.a0 != octeontx_svc_uuid[0] || res.a1 != octeontx_svc_uuid[1] ||
+	    res.a2 != octeontx_svc_uuid[2] || res.a3 != octeontx_svc_uuid[3])
+		return -EPERM;
+
+	return 0;
+}
+
 static ssize_t cn10k_rpram_info_read(struct file *f, char __user *user_buf,
 		size_t count, loff_t *off)
 {
@@ -141,6 +164,14 @@ static int __init cn10k_rpram_init(void)
 {
 	struct dentry *root, *entry;
 	struct device_node *parent, *node;
+	int ret;
+
+	ret = check_marvell_soc_cn10k();
+	if (ret) {
+		pr_info("%s: UIID SVC doesn't match Marvell CN10k.\n",
+			module_name(THIS_MODULE));
+		return ret;
+	}
 
 	parent = of_find_node_by_path("/reserved-memory");
 	if (!parent) {
-- 
2.31.1

